tests/lava/proc/io/test_buffer.py

```diff
@@ -0,0 +1,118 @@
+# Copyright (C) 2021-22 Intel Corporation
+# SPDX-License-Identifier: BSD-3-Clause
+# See: https://spdx.org/licenses/
+
+from typing import List
+import unittest
+import numpy as np
+from lava.magma.core.process.ports.ports import OutPort
+
+from lava.magma.core.process.variable import Var
+from lava.magma.core.run_configs import Loihi2SimCfg
+from lava.magma.core.run_conditions import RunSteps
+from lava.proc.io.injector import Injector
+from lava.proc.io.extractor import Extractor
+from lava.proc.lif.process import LIF
+from lava.proc.io.buffer import Buffer
+
+
+class TestBuffer(unittest.TestCase):
+    def test_create_buffer(self):
+        buffer = Buffer()
+        self.assertEqual(buffer.length, 100)
+        self.assertEqual(buffer.overflow, 'raise_error')
+
+    def test_connect_buffer(self):
+        buffer = Buffer()
+        # Test connecting to a Var
+        var_shape = (10,)
+        buffer_shape = (10, 100)
+        test_var = Var(shape=var_shape)
+        self.assertFalse(hasattr(buffer, 'Var1000'))
+        self.assertFalse(hasattr(buffer, 'Ref1000'))
+        buffer.connect(test_var)
+        self.assertTrue(hasattr(buffer, 'Var1000'))
+        self.assertTrue(hasattr(buffer, 'Ref1000'))
+        self.assertIn(buffer.Var1000, buffer.vars)
+        self.assertIn(buffer.Ref1000, buffer.ref_ports)
+        self.assertEqual(buffer.Ref1000.shape, var_shape)
+        self.assertEqual(buffer.Var1000.shape, buffer_shape)
+        # Test connecting to an OutPort
+        test_out_port = OutPort(shape=var_shape)
+        v = buffer.connect(test_out_port)
+        self.assertEqual(v.shape, buffer_shape)
+        self.assertTrue(hasattr(buffer, v.name))
+        self.assertEqual(len(list(buffer.in_ports)), 1)
+
+    def test_run_buffer(self):
+        num_steps = 10
+        injector = Injector(shape=(1,))
+        buffer = Buffer(length=10)
+        v0 = buffer.connect(injector.out_port)
+        buffer.create_runtime(run_cfg=Loihi2SimCfg())
+        buffer.run(condition=RunSteps(num_steps, blocking=False))
+        for t in range(num_steps):
+            injector.send(np.full((1,), t))
+        buffer.wait()
+        data = v0.get().flatten().astype(int).tolist()
+        buffer.stop()
+        self.assertSequenceEqual(data, [0, 1, 2, 3, 4, 5, 6, 7, 8, 9])
+
+    def test_buffer_with_2vars(self):
+        """Test to ensure that a Buffer can connect to multiple Vars."""
+        num_steps = 10
+        injector = Injector(shape=(1,))
+        lif = LIF(shape=(1,), du=1)
+        buffer = Buffer(length=10)
+        injector.out_port.connect(lif.a_in)
+        u = buffer.connect(lif.u)
```

Comment:
So this creates a `Var u` but does not place it inside the `buffer` object? I was somehow not expecting that.

Reply:
I can see how this is slightly confusing, because the use of dynamic port and var creation doesn't allow you to immediately see what's happening in PyBuffer, but the port and vars *are* added to the Buffer process and the corresponding PyPort and PyVars are added to the PyBuffer process model.

They can be accessed in the buffer.in_ports, buffer.out_ports, buffer.ref_ports, and buffer.vars collections, or they can be accessed directly with the use of a name as an attribute of the buffer. Unfortunately, I hit a small snag where I couldn't quite find a way to achieve user-specified names for those objects while also making it easy to intercept calls to the metaclass.getattr (necessary to convince the compiler to allow the process model to be built, but that's sort of a hack that could later be fixed in the compiler). Thus, the name of the var that will be used as a buffer is currently `VarXXXX` with the x's pulled from a incrementing sequence starting at 1000.

So, you could do:
```
buffer = Buffer()
u = buffer.connect(some_other_proc.some_port)
assert(u == buffer.Var1000)
```

I'll think a bit more on whether there's a way to let the user specify the name of the var, as I do find that more compelling than `Var1000`.